### DISCUSSION TITLE
[find_and_avoid] DDPG convergence issue

### DISCUSSION THREAD

**Discussion by polizogo67** [aidudezzz/deepworlds#15] [Q&A]
Hello, this is a more general Issue related to a robotics problem and not to this software directly. I just need help and I thought you might help. So my thesis is about obstacle avoidance and/or autonomous driving with RL. I have implemented DDPG and simple QL algorithms on webots but non of them seems to work. I also tried your ready to run example find and avoid but it did not seem to learn properly. Can you help me on how to train these algos ? cause I've tried them on Open AI Gym and they work fine.

If you want to communicate more efficiently with me this is my linked in and fb account. I am from Greece as two of the contributors are.
https://www.facebook.com/nathan.polizogo/
https://www.linkedin.com/in/athanassios-polizogopoulos-4839841b7/

**Comment by eakirtas** [comment]
Hello @NatePolizogo 

Of course there is not a single solution in your question. Indeed find-and-avoid example has some issues in convergence. From our experience, DDPG has convergence issues in generally contrary to others such as PPO. However, those issues comes form the continues nature of DDPG. Additionally, find target and avoid obstacle is not a simple task since has a big, continues observation space and continues actions. 

In this end, I suggest to experiment with the followings:
1) Is the implementation correct? You can take a look in [DDPG Stable Baseline](https://stable-baselines.readthedocs.io/en/master/modules/ddpg.html) or even use this exact implementation
2) Reward functions is crucial. Some experimentation would help
3) Is the neural network the appropriate one? Does the gradient flow through network during the backpropagation procedure? 
4) Does the literature propose any solutions?
5) It have been chosen the right observations?

Regarding to our implementation, I am pretty sure that there is a lot of space for improvements and we are constantly working in this direction. Any help will be appreciated.

BTW in which OpenAI gym problem did you use DDPG?

I have to close this issue since it's off topic but we can keep with thread to discuss any possible solution and even re-open in future if we come up with a solution. My colleague @tsampazk might offer some suggestions as well.

Cheers

**Comment by polizogo67** [comment]
Thank you very much for the instant reply. I'll try to implement PPO and SAC too and then I'll return to this discussion, I didn't knew that DDPG has convergence problems but I observed at all training I've done so far that I reached a maximum and then the performance was descending. 
To specify more my problem for now I use a pionner 3Dx with only the 6 front distance sensors but later they are going to be replaced by a camera and feature extractor network. Further more I would be happy if i could perform even a simpler task as obstacle avoidance only.
I want to try autonomous driving too but I cant find any racetrack on webots. Maybe I'll try make one in the future.
My reward function is my main concern to be the problem, that's why I tried deepwords. I've done some experimenting but I haven't achieved a good result.
I tried DDPG in LunarLanderContinuous-v2 and I implemented this from the youtube channel machine learning with phill. I think your implementation is based in his too. So I don't have many concerns that something is wrong with the DDPG.

Finally, I would be happy to help with anything, and I'll share any results and improvements, if any, with you.
After the completion of my thesis I could transfer the new additions to deepbots framework as for now I use a more simple approach with a single supervisor node, and help with further development.

**Comment by eakirtas** [comment]
Issue re-opened with a more general title if anyone wants to offer some ideas

**Comment by tsampazk** [comment]
Hey @NatePolizogo, first of all thank you for interest in deepbots!

I would like to add some general thoughts on this discussion that might or might not help:

1. I found that [creating your own robots from simple geometries](https://cyberbotics.com/doc/guide/tutorial-6-4-wheels-robot) might significantly improve performance so your experiments run faster, especially if you try to train several agents in parallel and thus include many Robot nodes in a simulation. 
2. In previous experiments we found that DDPG might require special care for its output, due to its continuous nature. It might saturate tanh-like functions or output high values with linear-like output activation functions. One solution would be to penalize high output values through the reward function. Another solution would be to add additional loss that penalizes high values on the output layer weights.
3. In some problems it might be required to progressively train an agent starting with simpler tasks and work your way up to the full task. This proved critical for us when working on similar problems of avoiding obstacles using distance sensors, as well as identifying and following a target through a camera sensor.
4. Webots includes some examples with city streets, e.g. `webotsInstallDir/projects/vehicles/worlds/city.wbt`, but i haven't personally used them.

**Comment by eakirtas** [comment]
> I didn't knew that DDPG has convergence problems

In most of the case convergence issues came from our poor understanding not from the algorithm. DDPG performs greatly in some cases and poorly in others. This might be a result of the chosen hyper parameters. Just for the record, I refer to you the  ["The problem with DDPG: understanding failures in deterministic environments with sparse rewards"](https://openreview.net/forum?id=HyxnH64KwS) paper

> I observed at all training I've done so far that I reached a maximum and then the performance was descending.

That is a common issue and it have been observed in our implementation as well. DDPG uses [Ornstein-Uhlenbec](https://journals.aps.org/pr/abstract/10.1103/PhysRev.36.823) process to generate noise in order to perform exploration. Tuning their hyper-parameters and decreasing the noise while training goes on+ might improves the model's performance ([here is the implementation](https://github.com/aidudezzz/deepworlds/blob/master/examples/find_and_avoid/controllers/supervisor/models/noise_generator.py)).    

As @tsampazk mentioned penalize model's output through reward is great place to start with (if you didn't).

Another issues we faced in this example was the poorly initialization. You can take a look in our DDPG implementation, [original paper](https://arxiv.org/pdf/1509.02971.pdf), [xevier](http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf) and [he](https://arxiv.org/pdf/1502.01852.pdf) initialization.

**Comment by polizogo67** [comment]
Thanks a lot, that a wide chunk of info, I'll make further investigation on these and try other algos too.

As I expected we use the exact same code for the noise.
I choose DDPG because I have a continuous enviroment on webots.

**Comment by tsampazk** [comment]
In case you meet a lot of difficulties with DDPG, you can try a continuous version of PPO. Generally, we found that PPO is a more beginner-friendly algorithm, that works well for many tasks.

[Here](https://github.com/nikhilbarhate99/PPO-PyTorch) you can find a minimal implementation of the continuous PPO algorithm with, using pytorch for the neural network backend, but this can easily be switched to your preferred backend. I have used it before and it works really well.

[Here](https://costa.sh/blog-the-32-implementation-details-of-ppo.html) is an interesting article regarding PPO implementation details and [here](https://arxiv.org/abs/2006.05990) you can find a great paper that i cannot recommend enough, that studies how various settings  affect the agent and convergence and provides helpful recommendations.
